alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/AbstractOrienting.kt

```diff
@@ -0,0 +1,417 @@
+         * find itself in a situation in which it is crossing a door but cannot locate
+         * itself inside any room. Such case is identified by CROSSING_DOOR: in this case,
+         * the pedestrian just proceeds towards the center of the room it wants to reach.
+         */
+        CROSSING_DOOR,
+        MOVING_TO_FINAL,
+        ARRIVED
+    }
+    private var state: State = State.START
+
+    private fun onStart() {
+        when {
+            environment.graph().nodeContaining(currentPosition) != null -> state = State.NEW_ROOM
+            /*
+             * If the pedestrian cannot locate itself inside any room (unusual
+             * condition), it tries to reach the closest door/passage in order
+             * to enter one. If this isn't possible, it simply won't move.
+             */
+            else -> moveToClosestDoor(ifNoDoorCanBeFound = { state = State.ARRIVED })
+        }
+    }
+
+    private fun inNewRoom() {
+        val newRoom = environment.graph().nodeContaining(currentPosition)
+            ?: throw IllegalStateException("pedestrian is not inside a room")
+        pedestrian.registerVisit(newRoom)
+        when {
+            ::nextRoom.isInitialized -> {
+                if (nextRoom.contains(currentPosition)) {
+                    currRoom = nextRoom
+                }
+                /*
+                 * We should have reached nextRoom, but we are in a different room (newRoom),
+                 * in such case we return back to the previous room (stored in currRoom), so
+                 * as to correctly follow our route.
+                 * In the future, something more sophisticated could be done (e.g. recomputing
+                 * the route).
+                 */
+                else {
+                    val prevRoom = currRoom
+                    currRoom = newRoom
+                    val doorsLeadingBack = environment.graph().outgoingEdgesOf(currRoom)
+                        .filter { environment.graph().getEdgeTarget(it) == prevRoom }
+                    moveToClosestDoor(doors = doorsLeadingBack)
+                    return
+                }
+            }
+            else -> currRoom = newRoom
+        }
+        environment.graph().destinationsWithin(currRoom).let { destinationsInSight ->
+            if (destinationsInSight.isNotEmpty()) {
+                route.clear()
+                subdestination = destinationsInSight.first()
+                state = State.MOVING_TO_FINAL
+                return
+            }
+        }
+        /*
+         * If a sub-destination of the route is in sight, updates the route removing
+         * all the sub-destination up to the one encountered.
+         */
+        if (route.isNotEmpty() && route.any { currRoom.contains(it.centroid) }) {
+            for (i in 0..route.indexOfLast { currRoom.contains(it.centroid) }) {
+                route.removeAt(0)
+            }
+        }
+        val rankings = if (route.isNotEmpty()) {
```

Comment:
route.takeUnless { it.isNotEmpty() }?.let { it.computeEdgeRankings(currRoom, route[0].centroid) }

looks cleaner IMHO

Reply:
and does not need an hybrid if with only one branch with curly brackets